pkg/kube/inject/watcher.go

```diff
@@ -95,23 +110,53 @@ func (w *fileWatcher) Run(stop <-chan struct{}) {
 	}
 }
 
+func (w *fileWatcher) Get() (*Config, string, error) {
+	return loadConfig(w.configFile, w.valuesFile)
+}
+
+func (w *fileWatcher) SetHandler(handler func(*Config, string)) {
+	w.handler = handler
+}
+
 // NewConfigMapWatcher creates a new Watcher for changes to the given ConfigMap.
-func NewConfigMapWatcher(client kube.Client, namespace, name, configKey, valuesKey string, callback func(*Config, string)) Watcher {
-	c := configmapwatcher.NewController(client, namespace, name, func(cm *v1.ConfigMap) {
+func NewConfigMapWatcher(client kube.Client, namespace, name, configKey, valuesKey string) Watcher {
+	w := &configMapWatcher{
+		client:    client,
+		namespace: namespace,
+		name:      name,
+		configKey: configKey,
+		valuesKey: valuesKey,
+	}
+	w.c = configmapwatcher.NewController(client, namespace, name, func(cm *v1.ConfigMap) {
 		sidecarConfig, valuesConfig, err := readConfigMap(cm, configKey, valuesKey)
 		if err != nil {
 			log.Warnf("failed to read injection config from ConfigMap: %v", err)
 			return
 		}
-		callback(sidecarConfig, valuesConfig)
+		if w.handler != nil {
+			w.handler(sidecarConfig, valuesConfig)
+		}
 	})
-	return &configMapWatcher{c: c}
+	return w
 }
 
 func (w *configMapWatcher) Run(stop <-chan struct{}) {
 	w.c.Run(stop)
 }
 
+func (w *configMapWatcher) Get() (*Config, string, error) {
+	cms := w.client.CoreV1().ConfigMaps(w.namespace)
+	cm, err := cms.Get(context.TODO(), w.name, metav1.GetOptions{})
+	if err != nil {
+		return nil, "", err
+	}
+	return readConfigMap(cm, w.configKey, w.valuesKey)
+}
+
+func (w *configMapWatcher) SetHandler(handler func(*Config, string)) {
```

Comment:
I think we need this to run prior to Run(), which I don't see happening here. if we want to call it after run, I would think we need to have this threadsafe (probably just a mutex) and possible immediately call the handler with .Get()? Not sure - but we need to make sure there isn't a race and that we don't drop the update

Reply:
SetHandler() happens in NewWebhook() and Run() happens after the webhook is created. But I can add a mutex anyway in case it's called elsewhere, since this isn't performance-critical

Reply:
The mutex alone doesn't fix it necessarily since you still miss the initial update. Maybe this is fine - and even with no mutex - but we should comment this on the SetHandler method and make sure we use it properly

Reply:
Ah I see. Ok yeah, ignore the commit I just made; I'll just add a comment to call this before Run